### PR TITLE
2D: Add Notes Provider, PrivateRoute component and update styles

### PR DESCRIPTION
**Pull request by StephanyPeralta**
This deliverable contains:

- Authentication validations for access to private routes
- Navigation routing through different views
- Create a new note with text and color
- Update a current note with text and color
- Search notes
- Archive notes
- Tests for private routes
- Authentication with real Backend with Firebase
- Add a 404 page
- Add Dark mode
- The app is responsive

### CODE REVIEW

File: src/components/NoteList/NoteList.component.tsx [hunk 1]
```diff
@@ -1,17 +1,25 @@
...
+function NoteList({ notes }: VideoListProps) {
   return (
     <NoteListWrapper>
-      <Note />
-      <Note />
-      <Note />
-      <Note />
-      <Note />
-      <Note />
+      {notes.map((note) => (
```
Comment (Raullg98) [inline src/components/NoteList/NoteList.component.tsx]: Can we move this notes to a higher level and receive them in this component as children?

File: src/pages/Home/Home.page.test.tsx [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+    const inputDescriptionNote = screen.getByText('test content');
+    const saveNoteButton = screen.getByTitle('Save');
+
+    userEvent.type(inputTitleNote, ' updated');
+    userEvent.type(inputDescriptionNote, ' updated');
+
+    await userEvent.click(saveNoteButton);
+
+    expect(screen.getByText('test title updated')).toBeInTheDocument();
+    expect(screen.getByText('test content updated')).toBeInTheDocument();
```
Comment (Raullg98) [inline src/pages/Home/Home.page.test.tsx:95]: Great job here!

File: src/pages/Home/Home.page.tsx [hunk 3]
```diff
@@ -2,16 +2,42 @@ import React from 'react';
 
 import AddNoteForm from '../../components/AddNoteForm';
 import NoteList from '../../components/NoteList';
-// import SessionForm from '../../components/SessionForm';
```
Comment (Raullg98) [inline src/pages/Home/Home.page.tsx:5]: Great!

File: src/components/Note/Note.component.tsx [hunk 4]
```diff
@@ -1,12 +1,153 @@
...
+
+  function handleArchiveNote() {
+    addArchiveNote(selectedNote);
+  }
+
+  function handleUnarchiveNote() {
+    removeArchiveNote(selectedNote);
+  }
+
+  useOnClickOutside(noteRef, () => {
```
Comment (Raullg98) [inline src/components/Note/Note.component.tsx:68]: I remember this hook from a previous project, it's great you added it here!

File: src/components/Note/Note.component.tsx [hunk 5]
```diff
@@ -1,12 +1,153 @@
...
+
+  const selectedNote = { id, title, description, bgColor };
+
+  const handleUpdateNote = () => {
+    setError('');
+    const title = noteTitle;
+    const description = noteDescription;
+    const bgColor = selectedColor;
+
+    if (title.trim().length > 0 && description.trim().length > 0) {
```
Comment (Raullg98) [inline src/components/Note/Note.component.tsx]: What if we add the opposite condition to do an early return. Check out this [example](https://forum.freecodecamp.org/t/the-return-early-pattern-explained-with-javascript-examples/19364).

File: src/components/Note/Note.component.tsx [hunk 6]
```diff
@@ -1,12 +1,153 @@
...
+  }, [editMode]);
 
-function Note() {
   return (
-    <NoteWrapper>
-      <input placeholder="Title" type="text" required />
-      <textarea rows={3} placeholder="Type to add a note..."></textarea>
+    <NoteWrapper ref={noteRef} style={{ backgroundColor: selectedColor }}>
+      <div className="note-content">
+        {!editMode ? (
```
Comment (Raullg98) [inline src/components/Note/Note.component.tsx:83]: What if we group all inputs using a React Frament <> ... </>. That way we can easily find the edit view and we can do the same for the read only version.

File: src/components/ProfileButton/ProfileButton.component.tsx [hunk 7]
```diff
@@ -1,35 +1,50 @@
...
 
+import { useAuth } from '../../providers/Auth';
 import useOnClickOutside from '../../hooks/useOnClickOutside';
 import { ProfileIconWrapper, Dropdown } from './ProfileButton.styled';
+import { toast } from 'react-toastify';
+import 'react-toastify/dist/ReactToastify.min.css';
 
 function ProfileButton() {
+  const { logout } = useAuth();
   const [openDropdown, setOpenDropdown] = useState(false);
```
Comment (Raullg98) [inline src/components/ProfileButton/ProfileButton.component.tsx:9]: Let's separate useState variables from hooks. It's easier to maintain in my opinion. Feel free to ignore this!

**Review by Raullg98 — CHANGES_REQUESTED** [review]
Good job! I requested a few changes, most of them are no big deal!

File: src/components/NoteList/NoteList.component.tsx (hunk 1 above)
Comment (StephanyPeralta) [reply src/components/NoteList/NoteList.component.tsx]: This change has been completed!

File: src/components/Note/Note.component.tsx (hunk 6 above)
Comment (StephanyPeralta) [reply src/components/Note/Note.component.tsx:83]: Yes, I agree... this change has been completed!

File: src/components/Note/Note.component.tsx (hunk 5 above)
Comment (StephanyPeralta) [reply src/components/Note/Note.component.tsx]: Thank you for the tip!

**Review by Raullg98 — APPROVED** [review]
Approved!

**Pull request merged**